90_mender_boot_grub.cfg

```diff
@@ -28,7 +28,10 @@ else
 fi
 
 if [ "${drop_to_grub_prompt}" = "no" ]; then
-    if linux ${mender_kernel_path}/${kernel_imagetype} root=${mender_root} ${bootargs}; then
+    probe --set devuuid --fs-uuid ${mender_grub_storage_device},gpt${mender_boot_part}
+    cryptomount ${mender_grub_storage_device},gpt${mender_boot_part}
+    mender_kernel_path="(crypto0)/boot"
+    if linux ${mender_kernel_path}/${kernel_imagetype} luks.name=${devuuid}=rootfs root=/dev/mapper/rootfs ${bootargs}; then
```

Comment:
Are FSUUIDs required for the `luks.name`? In general they are not guaranteed to be unique due to the fact that you can deploy any filesystem to the passive partition, even a previously "used" one.

This could even be an attack vector: If you deploy an artifact with a forged signature/checksum combination to a device, this will only be detected at the end of the download after the checksum has been calculated. The deployment will be aborted at that point, but the bytes, presumably including the FSUUID, will remain on disk, possibly confusing this logic and allowing booting from the attacker's image. Only partition indexes and PARTUUIDs can be trusted in this regard.

Reply:
Was thinking bit more about it. We can set UUID for rootA then different for rootB but with first update lets say from default state we overwrite rootB with mender artifact with UUID generated from rootA and the we set wrong UUID because we boot from rootB or am I missing something?

I think I'm missing point how PARTUUID work in general :)

Reply:
Ok, let's break it down:

* We leave PARTUUIDs out, as you say. These are not necessary for what we're looking at.

* `mender_boot_part` points to the currently active partition index, as in traditional Mender. It's either `2` or `3` for this example.

* Upon boot, the boot script does this:
  ```
  cryptsetup luksUUID --uuid  ${FIXED_ROOTFSA_UUID} /dev/mmcblk0p2
  cryptsetup luksUUID --uuid  ${FIXED_ROOTFSB_UUID} /dev/mmcblk0p3
  if ${mender_boot_part} = "2"; then
      LUKS_NAME = ${FIXED_ROOTFSA_UUID}
  else
      LUKS_NAME = ${FIXED_ROOTFSB_UUID}
  fi
  if linux ${mender_kernel_path}/${kernel_imagetype} luks.name=${LUKS_NAME}=rootfs root=/dev/mapper/rootfs ${bootargs}; then
      ...
  ```

In this case it's no problem to have the same UUIDs on each filesystem, because you're always resetting them to different values during boot, right?

Reply:
I'm not sure if grub have cryptsetup command available (only cryptomount IIRC)